In [38]:
!pip install pd

^C


In [56]:
import pandas as pd

# Подсказка по атрибутам train.csv

|Название атрибута|Тип данных|English desc|Русское описание|
|:------:|:--------:|:--------:|:--------:|
|row_id|(int64)|ID code for the row.|(int64) Код идентификатора строки.|
|timestamp|(int64)|the time in milliseconds between this user interaction and the first event completion from that user|время в миллисекундах между взаимодействием с пользователем и первым завершением события от этого пользователя|
|user_id|(int32)|ID code for the user|Код идентификатора пользователя|
|content_id|(int16)|ID code for the user interaction|Код идентификатора взаимодействия пользователя|
|content_type_id|(int8)|0 if the event was a question being posed to the user, 1 if the event was the user watching a lecture|0, если событие было вопросом, заданным пользователю, 1, если событием было то что пользователь смотрел лекцию|
|task_container_id|(int16)|Id code for the batch of questions or lectures. For example, a user might see three questions in a row before seeing the explanations for any of them. Those three would all share a task_container_id|Код идентификатора для пакета вопросов или лекций. Например, пользователь может увидеть три вопроса подряд, прежде чем увидеть объяснения для любого из них. Все эти три вопроса будут иметь общий task_container_id.|
|user_answer|(int8)|the user's answer to the question, if any. Read -1 as null, for lectures|ответ пользователя на вопрос, если таковой имеется. Читать -1 как null, для лекций|
|answered_correctly|(int8)|if the user responded correctly. Read -1 as null, for lectures|если пользователь ответил правильно. Читать -1 как null, для лекций|
|prior_question_elapsed_time|(float32)|The average time in milliseconds it took a user to answer each question in the previous question bundle, ignoring any lectures in between. Is null for a user's first question bundle or lecture.Note that the time is the average time a user took to solve each question in the previous bundle|Среднее время в миллисекундах, которое потребовалось пользователю для ответа на каждый вопрос в предыдущем пакете вопросов, игнорируя любые лекции между ними. Имеет значение null для первого пакета вопросов или лекции пользователя. Обратите внимание, что время — это среднее время, которое потребовалось пользователю для решения каждого вопроса в предыдущем пакете|
|prior_question_had_explanation|(bool)|Whether or not the user saw an explanation and the correct response(s) after answering the previous question bundle, ignoring any lectures in between. The value is shared across a single question bundle, and is null for a user's first question bundle or lecture. Typically the first several questions a user sees were part of an onboarding diagnostic test where they did not get any feedback.|Увидел ли пользователь объяснение и правильный ответ(ы) после ответа на предыдущий пакет вопросов, игнорируя любые лекции между ними. Значение является общим для одного пакета вопросов и имеет значение null для первого пакета вопросов или лекции пользователя. Обычно первые несколько вопросов, которые видит пользователь, были частью диагностического теста при вступлении, в ходе которого он не получал никакой обратной связи|



In [57]:
df_train = pd.read_csv('data/train.csv', nrows=1_000_000)

In [58]:
df_train

row_id  timestamp   user_id  content_id  content_type_id  \
0            0          0       115        5692                0   
1            1      56943       115        5716                0   
2            2     118363       115         128                0   
3            3     131167       115        7860                0   
4            4     137965       115        7922                0   
...        ...        ...       ...         ...              ...   
999995  999995   26482248  20949024        8803                0   
999996  999996   26516686  20949024        4664                0   
999997  999997   26537967  20949024        4108                0   
999998  999998   26590240  20949024        5014                0   
999999  999999   26605316  20949024        9828                0   

        task_container_id  user_answer  answered_correctly  \
0                       1            3                   1   
1                       2            2                   1   
2                       0            0                   1   
3                       3            0                   1   
4                       4            1                   1   
...                   ...          ...                 ...   
999995                 29            1                   1   
999996                 30            3                   1   
999997                 31            1                   0   
999998                 32            3                   0   
999999                 33            0                   0   

        prior_question_elapsed_time prior_question_had_explanation  
0                               NaN                            NaN  
1                           37000.0                          False  
2                           55000.0                          False  
3                           19000.0                          False  
4                           11000.0                          False  
...                             ...                            ...  
999995                      14000.0                           True  
999996                      17000.0                           True  
999997                      18000.0                           True  
999998                       6000.0                           True  
999999                      46000.0                           True  

[1000000 rows x 10 columns]

In [59]:
df_train.describe()

row_id     timestamp       user_id      content_id  \
count  1000000.000000  1.000000e+06  1.000000e+06  1000000.000000   
mean    499999.500000  7.344311e+09  1.017255e+07     5232.691523   
std     288675.278932  1.058693e+10  6.029234e+06     3882.323419   
min          0.000000  0.000000e+00  1.150000e+02        0.000000   
25%     249999.750000  5.211669e+08  4.702585e+06     2057.000000   
50%     499999.500000  2.824811e+09  9.678259e+06     5052.000000   
75%     749999.250000  1.010186e+10  1.556872e+07     7421.000000   
max     999999.000000  7.809200e+10  2.094902e+07    32736.000000   

       content_type_id  task_container_id     user_answer  answered_correctly  \
count   1000000.000000     1000000.000000  1000000.000000      1000000.000000   
mean          0.019907         806.105689        1.374784            0.617504   
std           0.139681        1025.887530        1.194168            0.525364   
min           0.000000           0.000000       -1.000000           -1.000000   
25%           0.000000         109.000000        0.000000            0.000000   
50%           0.000000         391.000000        1.000000            1.000000   
75%           0.000000        1112.000000        3.000000            1.000000   
max           1.000000        7739.000000        3.000000            1.000000   

       prior_question_elapsed_time  
count                976277.000000  
mean                  25302.962693  
std                   19710.310227  
min                       0.000000  
25%                   16000.000000  
50%                   21000.000000  
75%                   29666.000000  
max                  300000.000000

In [60]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000000 entries, 0 to 999999
Data columns (total 10 columns):
 #   Column                          Non-Null Count    Dtype  
---  ------                          --------------    -----  
 0   row_id                          1000000 non-null  int64  
 1   timestamp                       1000000 non-null  int64  
 2   user_id                         1000000 non-null  int64  
 3   content_id                      1000000 non-null  int64  
 4   content_type_id                 1000000 non-null  int64  
 5   task_container_id               1000000 non-null  int64  
 6   user_answer                     1000000 non-null  int64  
 7   answered_correctly              1000000 non-null  int64  
 8   prior_question_elapsed_time     976277 non-null   float64
 9   prior_question_had_explanation  996184 non-null   object 
dtypes: float64(1), int64(8), object(1)
memory usage: 76.3+ MB


Заменим -1 в столбцах `answered_correctly` на NaN чтобы подсчить какой процент правильных ответов был дан студентами на все вопросы, относитьельно нашей выборки в 1_000_000 строк

In [61]:
df_train.loc[(df_train['answered_correctly'] == -1), 'answered_correctly'] = pd.NA
df_train['answered_correctly'].mean()

0.6503576701394663

Отсюда можно сделать вывод что **65% вопросов**, были отвечены студентами правильно

# Подсказка по атрибутам lectures.csv

|Название атрибута|English desc|Русское описание|
|:------:|:--------:|:--------:|
|lecture_id|foreign key for the train/test content_id column, when the content type is lecture (1)|внешний ключ для столбца content_id обучения/тестирования, где тип контента — лекция (1)|
|part|top level category code for the lecture|код категории верхнего уровня для лекции|
|tag|one tag codes for the lecture. The meaning of the tags will not be provided, but these codes are sufficient for clustering the lectures together|один код тега для лекции. Значение тегов не будет предоставлено, но этих кодов достаточно для кластеризации лекций вместе|
|type_of|brief description of the core purpose of the lecture|краткое описание основной цели лекции|

In [62]:
df_lectures = pd.read_csv('data/lectures.csv')

In [63]:
df_lectures

lecture_id  tag  part           type_of
0            89  159     5           concept
1           100   70     1           concept
2           185   45     6           concept
3           192   79     5  solving question
4           317  156     5  solving question
..          ...  ...   ...               ...
413       32535    8     5  solving question
414       32570  113     3  solving question
415       32604   24     6           concept
416       32625  142     2           concept
417       32736   82     3           concept

[418 rows x 4 columns]

In [64]:
df_lectures.describe()

lecture_id         tag        part
count    418.000000  418.000000  418.000000
mean   16983.401914   94.480861    4.267943
std     9426.164660   53.586487    1.872424
min       89.000000    0.000000    1.000000
25%     9026.250000   50.250000    2.000000
50%    17161.500000   94.500000    5.000000
75%    24906.250000  140.000000    6.000000
max    32736.000000  187.000000    7.000000

In [65]:
df_lectures.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   lecture_id  418 non-null    int64 
 1   tag         418 non-null    int64 
 2   part        418 non-null    int64 
 3   type_of     418 non-null    object
dtypes: int64(3), object(1)
memory usage: 13.2+ KB


# Подсказка по атрибутам questions.csv

|Название атрибута|English desc|Русское описание|
|:------:|:--------:|:--------:|
|question_id|foreign key for the train/test content_id column, when the content type is question (0)|внешний ключ для столбца content_id обучения/тестирования, когда тип контента — вопрос (0)|
|bundle_id|code for which questions are served together|код, для которого вопросы обслуживаются вместе|
|correct_answer|the answer to the question. Can be compared with the train user_answer column to check if the user was right|ответ на вопрос. Можно сравнить со столбцом user_answer train.csv, чтобы проверить, был ли пользователь прав|
|part|the relevant section of the TOEIC test|соответствующий раздел теста TOEIC|
|tags|one or more detailed tag codes for the question. The meaning of the tags will not be provided, but these codes are sufficient for clustering the questions together|один или несколько подробных кодов тегов для вопроса. Значение тегов не будет предоставлено, но этих кодов достаточно для кластеризации вопросов вместе|


In [66]:
df_questions = pd.read_csv('data/questions.csv')

In [67]:
df_questions

question_id  bundle_id  correct_answer  part            tags
0                0          0               0     1   51 131 162 38
1                1          1               1     1       131 36 81
2                2          2               0     1  131 101 162 92
3                3          3               0     1  131 149 162 29
4                4          4               3     1    131 5 162 38
...            ...        ...             ...   ...             ...
13518        13518      13518               3     5              14
13519        13519      13519               3     5               8
13520        13520      13520               2     5              73
13521        13521      13521               0     5             125
13522        13522      13522               3     5              55

[13523 rows x 5 columns]

In [68]:
df_questions.describe()

question_id     bundle_id  correct_answer          part
count  13523.00000  13523.000000    13523.000000  13523.000000
mean    6761.00000   6760.510907        1.455298      4.264956
std     3903.89818   3903.857783        1.149707      1.652553
min        0.00000      0.000000        0.000000      1.000000
25%     3380.50000   3379.500000        0.000000      3.000000
50%     6761.00000   6761.000000        1.000000      5.000000
75%    10141.50000  10140.000000        3.000000      5.000000
max    13522.00000  13522.000000        3.000000      7.000000

In [69]:
df_questions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13523 entries, 0 to 13522
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   question_id     13523 non-null  int64 
 1   bundle_id       13523 non-null  int64 
 2   correct_answer  13523 non-null  int64 
 3   part            13523 non-null  int64 
 4   tags            13522 non-null  object
dtypes: int64(4), object(1)
memory usage: 528.4+ KB


Проверим зависимость времени за которые студент ответил на вопрос и корректностью ответа

In [70]:
# Получаем выборку студентов, которые отвечали правильно
correct_answer_filt = df_train[(df_train['content_type_id'] == 0) & (df_train['answered_correctly'] == 1)]

# Получаем выборку студентов, которые отвечали правильно
ancorrect_answer_filt = df_train[(df_train['content_type_id'] == 0) & (df_train['answered_correctly'] == 0)]

# Выводим основные статистические свойства, получившихся выборок 
display(correct_answer_filt['prior_question_elapsed_time'].describe())
display(ancorrect_answer_filt['prior_question_elapsed_time'].describe())

count    634814.000000
mean      25349.965658
std       19501.844786
min           0.000000
25%       16000.000000
50%       21000.000000
75%       29250.000000
max      300000.000000
Name: prior_question_elapsed_time, dtype: float64

count    341463.000000
mean      25215.579462
std       20091.857457
min           0.000000
25%       15000.000000
50%       21000.000000
75%       30000.000000
max      300000.000000
Name: prior_question_elapsed_time, dtype: float64


Из полученных результатов видно, что в **среднем**, чтобы ответить правильно студенты необходимо потратить **больше времени**, чем студентам, которые не знали правильного ответа и отвечали неправильно (но также, нельзя исключать человеческий фактор и событие, при котором студент мог знать ответ, но по какой то причине оветил неправильно)

Также, можно заметить смещение **квантилей 25% и 75%** у полученных выборок.

Можно предположить, что смещение в большую сторону **25% квантиля** у студентов, которые отвечали на вопросы правильно, обусловлено более вдумчивым подходом к ответу на несложные вопросы (Возьмем градацию сложности вопроса по времение ответа на него, чем сложнее вопрос тем больше времени на него тратит студент).
Тогда как студенты, которые ответили неправильно, скорее всего под влиянием стресса и легкой паники из за незнания ответа, старались как можно быстрее "избавиться" от этого чувства, по средствам ответа на вопрос, пускай даже и неправильного, так сказать "На удачу"

**Смещение 75% квартиля**, в меньшую сторону у студентов, отвечавших правильно, связываю с тем что они более быстро справлялись с разбором и понимаем сложных вопросов 
благодаря этому отвечали немного быстрее, чем вторая группа студентов.

Теперь проверим 

In [71]:
content_type_id_true = df_train[(df_train['content_type_id'] == 1)]
students_read_the_lecture = df_train[(df_train['user_id'].isin(content_type_id_true['user_id'])) & (df_train['content_type_id'] == 0)]
students_didnt_read_the_lecture = df_train[(~df_train['user_id'].isin(content_type_id_true['user_id']))]
display('Количество студентов в 1 выборке', students_read_the_lecture['user_id'].nunique())
display(students_read_the_lecture.describe())
display('Количество студентов в 2 выборке', students_didnt_read_the_lecture['user_id'].nunique())
display(students_didnt_read_the_lecture.describe())

'Количество студентов в 1 выборке'

1475

row_id     timestamp       user_id     content_id  \
count  859485.000000  8.594850e+05  8.594850e+05  859485.000000   
mean   494214.570923  8.038416e+09  1.004941e+07    5046.012378   
std    287679.148552  1.086712e+10  6.006763e+06    3332.531082   
min        76.000000  0.000000e+00  2.746000e+03       0.000000   
25%    245273.000000  8.570842e+08  4.663746e+06    1924.000000   
50%    489053.000000  3.576460e+09  9.418512e+06    5026.000000   
75%    744081.000000  1.125718e+10  1.544139e+07    7405.000000   
max    999900.000000  7.809200e+10  2.093825e+07   13522.000000   

       content_type_id  task_container_id    user_answer  answered_correctly  \
count         859485.0      859485.000000  859485.000000       859485.000000   
mean               0.0         896.557122       1.424036            0.660920   
std                0.0        1060.043433       1.156942            0.473397   
min                0.0           0.000000       0.000000            0.000000   
25%                0.0         163.000000       0.000000            0.000000   
50%                0.0         479.000000       1.000000            1.000000   
75%                0.0        1249.000000       3.000000            1.000000   
max                0.0        7739.000000       3.000000            1.000000   

       prior_question_elapsed_time  
count                858014.000000  
mean                  25389.942145  
std                   19764.357762  
min                       0.000000  
25%                   16000.000000  
50%                   21000.000000  
75%                   30000.000000  
max                  300000.000000

'Количество студентов в 2 выборке'

2349

row_id     timestamp       user_id     content_id  \
count  120608.000000  1.206080e+05  1.206080e+05  120608.000000   
mean   539883.595334  2.306701e+09  1.102140e+07    4674.042576   
std    292910.966712  6.158061e+09  6.126130e+06    2923.455243   
min         0.000000  0.000000e+00  1.150000e+02       0.000000   
25%    294330.750000  4.458595e+05  5.493032e+06    2184.000000   
50%    569319.500000  5.510879e+07  1.160998e+07    4696.000000   
75%    811079.250000  1.097629e+09  1.646714e+07    6789.000000   
max    999999.000000  7.647396e+10  2.094902e+07   13522.000000   

       content_type_id  task_container_id    user_answer  answered_correctly  \
count         120608.0      120608.000000  120608.000000       120608.000000   
mean               0.0         176.998425       1.415768            0.575086   
std                0.0         398.979690       1.155565            0.494332   
min                0.0           0.000000       0.000000            0.000000   
25%                0.0           9.000000       0.000000            0.000000   
50%                0.0          23.000000       1.000000            1.000000   
75%                0.0         114.000000       3.000000            1.000000   
max                0.0        2584.000000       3.000000            1.000000   

       prior_question_elapsed_time  
count                118263.000000  
mean                  24671.915045  
std                   19302.007580  
min                       0.000000  
25%                   15666.000000  
50%                   20000.000000  
75%                   28500.000000  
max                  300000.000000

На основе получившихся данных можно получить предварительный вывод, что прочтение студентами лекций положительно сказывается на их ответах на вопросы.

**66% правильно** отвеченных вопросов и **859485** в принципе отвеченных вопросов было сделано **1475 студентами**.
Тогда как, **57.5% правильно** отвеченных вопросов и всего лишь **120608** в принципе отвеченных вопросов было сделано **2349 студентами**.

То есть в **1,6** раза меньшее количество студентов, ответило **в 7** раз больше вопросов с процентом правильных оветов **66%**.